include/hsprov_store.h

```diff
@@ -0,0 +1,132 @@
+/**
+ * @file hsprov_store.h class definition of a cassandra-backed store.
+ *
+ * Copyright (C) Metaswitch Networks 2017
+ * If license terms are provided to you in a COPYING file in the root directory
+ * of the source code repository by which you are accessing this code, then
+ * the license outlined in that COPYING file applies to your use.
+ * Otherwise no rights are granted except for those provided to you by
+ * Metaswitch Networks in a separate written agreement.
+ */
+#ifndef HSPROV_STORE_H__
+#define HSPROV_STORE_H__
+
+#include "cassandra_store.h"
+#include "reg_state.h"
+#include "charging_addresses.h"
+#include "authvector.h"
+
+class HsProvStore : public CassandraStore::Store
+{
+public:
+  virtual ~HsProvStore();
+
+  /// @return the singleton cache instance.
+  static inline HsProvStore* get_instance() { return INSTANCE; }
+
+private:
+  // Singleton variables.
+  static HsProvStore* INSTANCE;
+  static HsProvStore DEFAULT_INSTANCE;
+
+protected:
+  // The constructors and assignment operation are protected to prevent multiple
+  // instances of the class from being created.
+  HsProvStore();
+  HsProvStore(HsProvStore const&);
+  void operator=(HsProvStore const&);
+
+public:
+  //
+  // Operations
+  //
+
+  class GetRegData : public CassandraStore::HAOperation
+  {
+  public:
+    /// Get the IMS subscription XML for a public identity.
+    ///
+    /// @param public_id the public identity.
+    GetRegData(const std::string& public_id);
+    virtual ~GetRegData();
+
+    /// Access the result of the request.
+    ///
+    /// @param xml the IMS subscription XML document.
+    virtual void get_xml(std::string& xml);
```

Comment:
these are actually the base classes, and they're virtual to alllow them to be mocked out in UTs